File: src/com/daveme/intellij/chocolateCakePHP/navigation/ViewHelperGotoDeclarationHandler.java
package com.daveme.intellij.chocolateCakePHP.navigation;

import com.daveme.intellij.chocolateCakePHP.util.CakeUtil;
import com.daveme.intellij.chocolateCakePHP.util.PsiUtil;
import com.intellij.codeInsight.navigation.actions.GotoDeclarationHandler;
import com.intellij.openapi.actionSystem.DataContext;
import com.intellij.openapi.editor.Editor;
import com.intellij.patterns.PlatformPatterns;
import com.intellij.psi.PsiElement;
import com.jetbrains.php.lang.psi.elements.FieldReference;
import com.jetbrains.php.lang.psi.elements.PhpClass;
import com.jetbrains.php.lang.psi.elements.PhpExpression;
import com.jetbrains.php.lang.psi.resolve.types.PhpType;
import org.jetbrains.annotations.Nullable;

import java.util.Collection;

public class ViewHelperGotoDeclarationHandler implements GotoDeclarationHandler {
    @Nullable
    @Override
    public PsiElement[] getGotoDeclarationTargets(@Nullable PsiElement psiElement, int i, Editor editor) {
        if (psiElement == null) {
            return PsiElement.EMPTY_ARRAY;
        }
        if (!PlatformPatterns.psiElement().withParent(FieldReference.class).accepts(psiElement)) {
            return PsiElement.EMPTY_ARRAY;
        }
        PsiElement parent = psiElement.getParent();
        if (parent == null) {
            return PsiElement.EMPTY_ARRAY;
        }
        FieldReference fieldReference = (FieldReference)parent;
        String fieldName = fieldReference.getName();
        if (fieldName == null) {
            return PsiElement.EMPTY_ARRAY;
        }
        PhpExpression classReference = fieldReference.getClassReference();
        if (classReference == null) {
            return null;
        }
        PhpType types = classReference.getType();
        String fieldReferenceName = fieldReference.getName();
        if (fieldReferenceName == null) {
            return null;
        }
        if (!Character.isUpperCase(fieldReferenceName.charAt(0))) {
            return null;
        }
        for (String type : types.getTypes()) {
            if (type.contains("#Vthis")) {
                Collection<PhpClass> classes = PsiUtil.getViewHelperClasses(fieldName, psiElement.getProject());
                return classes.toArray(new PsiElement[classes.size()]);
            }
        }
        return PsiElement.EMPTY_ARRAY;
    }

    @Nullable
    @Override
    public String getActionText(DataContext dataContext) {
        return null;
    }
}

File: src/com/daveme/intellij/chocolateCakePHP/navigation/ViewHelperGotoDeclarationHandler.kt
package com.daveme.intellij.chocolateCakePHP.navigation

import com.daveme.intellij.chocolateCakePHP.util.getClassesForViewHelper
import com.daveme.intellij.chocolateCakePHP.util.startsWithUppercaseCharacter
import com.intellij.codeInsight.navigation.actions.GotoDeclarationHandler
import com.intellij.openapi.actionSystem.DataContext
import com.intellij.openapi.editor.Editor
import com.intellij.patterns.PlatformPatterns
import com.intellij.psi.PsiElement
import com.jetbrains.php.lang.psi.elements.FieldReference

class ViewHelperGotoDeclarationHandler : GotoDeclarationHandler {
    override fun getGotoDeclarationTargets(psiElement: PsiElement?, i: Int, editor: Editor): Array<PsiElement>? {
        if (psiElement == null) {
            return PsiElement.EMPTY_ARRAY
        }
        if (!PlatformPatterns.psiElement().withParent(FieldReference::class.java).accepts(psiElement)) {
            return PsiElement.EMPTY_ARRAY
        }
        val parent = psiElement.parent ?: return PsiElement.EMPTY_ARRAY
        val fieldReference = parent as FieldReference
        val fieldName = fieldReference.name ?: return PsiElement.EMPTY_ARRAY
        val classReference = fieldReference.classReference ?: return PsiElement.EMPTY_ARRAY
        val fieldReferenceName = fieldReference.name
        if (!fieldReferenceName.startsWithUppercaseCharacter()) {
            return PsiElement.EMPTY_ARRAY
        }
        if (classReference.text == "\$this") {
            return getClassesForViewHelper(psiElement.project, fieldName).toTypedArray()
        }
        return PsiElement.EMPTY_ARRAY
    }

    override fun getActionText(dataContext: DataContext): String? {
        return null
    }
}